packages/manager/apps/pci-volume-snapshot/src/api/hooks/useSnapshots.ts

```diff
@@ -0,0 +1,68 @@
+import { useMemo } from 'react';
+import { useQuery, useQueries } from '@tanstack/react-query';
+import { ColumnSort, PaginationState } from '@ovh-ux/manager-react-components';
+import { applyFilters, Filter } from '@ovh-ux/manager-core-api';
+import { getSnapshots, getVolume, TSnapshot, TVolume } from '../data/snapshots';
+import { paginateResults, sortResults } from '@/helpers';
+
+export type TVolumeSnapshot = TSnapshot & {
+  volume?: TVolume | null;
+};
+
+export const useAllSnapshots = (projectId: string) =>
+  useQuery({
+    queryKey: ['snapshots', projectId],
+    enabled: !!projectId,
+    queryFn: () => getSnapshots(projectId),
+  });
+
+export const useVolumeSnapshots = (projectId: string) => {
+  const { data: snapshots, isPending, isLoading } = useAllSnapshots(projectId);
+
+  const volumeIds = snapshots?.map((volume) => volume.volumeId);
+  const uniqueVolumeIds = [...new Set(volumeIds)];
+
+  return useQueries({
+    queries: uniqueVolumeIds?.map((volumeId) => ({
+      queryKey: ['snapshots', projectId, 'volume', volumeId],
+      queryFn: () => getVolume(projectId, volumeId),
+    })),
+    combine: (volumes) => ({
+      isPending: isPending || volumes.some((result) => result.isPending),
+      isLoading: isLoading || volumes.some((result) => result.isLoading),
+      error: volumes.find((result) => result.error),
+      data: (snapshots || []).map((snapshot) => ({
+        ...snapshot,
+        volume: volumes.find(({ data }) => data?.id === snapshot.volumeId)
+          ?.data,
+      })),
+    }),
+  });
+};
+
+export const usePaginatedVolumeSnapshot = (
```

Comment:
This should could be a plain `select` function given as parameter to the `useVolumeSnapshots` hook.
This way we could extend the initial hook with agnostic computing behavior without rewriting some hooks that have no direct value :)

Reply:
I like the idea but it must be a hook so that when filters change, the list is updated (without any server roundtrip).

If I pass a plain select function to `useVolumeSnapshots` I loose this capability.
Data won't update when filters update because `useVolumeSnapshots` has no idea about `pagination`, `sorting`, `filters`.

Or that means, in the view, we have to create a `useMemo` that calls `useVolumeSnapshots` as soon as `pagination`, `sorting` or `filters` changes.
But that would only move a part of the code to the view.

What do you think?
Did you have any other solution in mind?

Reply:
Yes, it's pretty hard to implement it this way because of the current architecture of both the app (react-query) and the API.

We could achieve this using functional composition (currification) but it would complicate a lot the existing code, so IMO the best solution would be to redesign API and handle sort, filter & pagination on backend-side, so data will be filtered in a reactive way with no extra logic on the view :)

But agree that, for now, we could go on with the hook, thanks for your analysis :)